### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    this.errors = errors;
+  }
+}
+
+export const validateTodoCreate = (req: Request, res: Response, next: NextFunction) => {
+  const errors: ValidationError[] = [];
+  const { title, description } = req.body;
+
+  // Validate title
+  if (!title) {
```
Comment (dandeliongold) [inline src/middleware/validation.ts:24]: In `validateTodoCreate` the title check starts with `if (!title)`. This will treat `''`, `0`, and `null` similarly; prefer distinguishing missing vs empty string for clearer errors. For example, use `if (title === undefined)` for missing and `else if (typeof title === 'string' && title.trim().length === 0)` for empty strings.

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title must be 100 characters or less' });
+  } else if (/<[^>]*>/g.test(title)) {
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
```
Comment (dandeliongold) [inline src/middleware/validation.ts:40]: The description HTML-detection regex is `/<.*>/g` which is greedy and different from the title check. This can allow some tags through or match across content. Use a single safe pattern (e.g. `/<[^>]*>/g`) or better, use a sanitizer library. Also note the inline comment already indicating this as a bug — please fix to keep behavior consistent.

File: src/middleware/validation.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<[^>]*>/g.test(description)) {
+      errors.push({ field: 'description', message: 'Description cannot contain HTML tags' });
+    }
+  }
+
+  // Validate completed if provided
+  if (completed !== undefined && typeof completed !== 'boolean') {
+    errors.push({ field: 'completed', message: 'Completed must be a boolean', value: completed });
+  }
```
Comment (dandeliongold) [inline src/middleware/validation.ts:88]: ID validation uses `/^\d+$/` which accepts `"0"`. If IDs are 1-based, change to `/^[1-9]\d*$/` to reject zero. If `0` should be allowed, add a comment explaining why.

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
+import request from 'supertest';
+import express from 'express';
+import { validateTodoCreate, validateTodoUpdate, validateTodoId } from './validation';
+
+const createTestApp = (middleware: any) => {
+  const app = express();
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:6]: `validateTodoUpdate` is imported but not tested. Add test cases for update validation (optional title/description trimming, `completed` type enforcement, and HTML tag detection). This will ensure parity with `validateTodoCreate` tests.

File: src/middleware/validation.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        .post('/')
+        .send({ title: '   ' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot be empty'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:49]: There's a comment saying a test is missing for description HTML validation — please add a test that submits a `description` containing tags and asserts the appropriate error, mirroring the title test for HTML tags.

File: src/routes/todos.ts [hunk 6]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
 });
 
-router.get('/:id', (req: Request, res: Response) => {
+router.get('/:id', validateTodoId, (req: Request, res: Response) => {
   const { id } = req.params;
   const todo = todoStore.getById(id);
   
   if (!todo) {
-    return res.status(404).json({ error: 'Todo not found' });
+    return res.status(404).json({
```
Comment (dandeliongold) [inline src/routes/todos.ts:18]: `todoStore.create` is called without a try/catch. If the store can throw, this will crash the server. Wrap `todoStore.create` in try/catch and return a `500` with a consistent error shape, or rely on a centralized error handler and document expectations.

File: package.json [hunk 7]
```diff
@@ -12,7 +12,13 @@
...
   },
-  "keywords": ["typescript", "express", "todo", "api", "demo"],
+  "keywords": [
+    "typescript",
+    "express",
+    "todo",
+    "api",
+    "demo"
+  ],
   "author": "",
```
Comment (dandeliongold) [inline package.json:22]: `supertest` and `@types/supertest` are listed under `dependencies`. These are test-only; please move them to `devDependencies` so production installs don't include test tooling.

**Review by dandeliongold — CHANGES_REQUESTED** [review]
Thanks — this PR adds important validation and tests. I left several inline comments about correctness, consistency, and test coverage. Summary of required changes:

- Unify HTML tag detection regex (avoid `/<.*>/g`) and fix the noted bug.
- Enforce positive non-zero IDs (or document why `0` is valid).
- Distinguish missing vs empty `title` for clearer error messages.
- Add try/catch around `todoStore.create` or ensure centralized error handling.
- Add tests for `validateTodoUpdate`, description HTML validation, and the ID `"0"` edge case.
- Move `supertest` and its types to `devDependencies`.

Once those are addressed I can re-review and approve.
